internal/pkg/agent/cmd/container_init_linux.go

```diff
@@ -0,0 +1,313 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"kernel.org/pub/linux/libs/security/libcap/cap"
+
+	"github.com/elastic/elastic-agent/internal/pkg/cli"
+	"github.com/elastic/elastic-agent/pkg/utils"
+)
+
+var (
+	// for unit-testing
+	capBound   = cap.GetBound
+	capGetFile = cap.GetFile
+	capProc    = cap.GetProc
+)
+
+// initContainer applies the following container initialisation steps:
+// - set any missing capabilities of Effective set based on the Bounding set at elastic-agent binary
+// - raise capabilities of the Ambient set to match the Effective set
+// - chown all agent-related paths if DAC_OVERRIDE capability is not in the Effective set
+// If new binary capabilities are set then the returned cmd will be not nil. Note that it is up to caller to invoke
+// the returned cmd and spawn an agent instance with all the capabilities.
+func initContainer(streams *cli.IOStreams) (*exec.Cmd, error) {
+	isRoot, _ := utils.HasRoot()
+	if !skipFileCapabilities && !isRoot {
+		executable, err := os.Executable()
+		if err != nil {
+			return nil, err
+		}
+
+		logInfo(streams, "agent container initialisation - file capabilities")
+		updated, err := updateFileCapsFromBoundingSet(executable)
+		if err != nil {
+			return nil, err
+		}
+
+		if updated {
+			// new capabilities were added thus we need to re-exec agent to pick them up
+			var args []string
+			if len(os.Args) > 1 {
+				args = append(args, os.Args[1:]...)
+			}
+			// add skipFileCapabilitiesFlag flag to skip reapplying the file capabilities
+			args = append(args, fmt.Sprintf("--%s", skipFileCapabilitiesFlag))
+
+			cmd := exec.Command(executable, args...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			cmd.Env = os.Environ()
+			return cmd, nil
```

Comment:
This isn't really a re-exec, this is running it as a subprocess for the live of the Elastic Agent. Is that what really should be done or should it be a true re-exec?

We have the ability to do a true re-exec, that is how the Elastic Agent handles self-upgrading.

Reply:
fantastic! true re-exec is the way to go, can you please point me to the appropriate line in the code so I can try to incorporate it? 🙂

Reply:
oh unix.Exec 🥲 ok I am gonna try to incorporate it

Reply:
thanks for the hint/tip here, I switched to unix.Exec, feel free to unresolve if you spot anything else